src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -230,41 +232,98 @@ private static (SyntaxTriviaList sameLine, SyntaxTriviaList endOfLine) SplitTrai
             ISyntaxFactsService syntaxFactsService,
             CancellationToken cancellationToken)
         {
-            foreach (var nodeOrToken in declarationSyntax.Parent.ChildNodesAndTokens().Reverse())
+            var firstVariableUsage = new Dictionary<ILocalSymbol, int>();
+            var lastVariableUsage = new Dictionary<ILocalSymbol, int>();
+            var localVariables = new List<ILocalSymbol> { localVariable };
+
+            // Take all the statements after the trigger variable's declaration.
+            var statementsAfterDeclaration = declarationSyntax.Parent.ChildNodesAndTokens()
+                .Select(nodeOrToken => nodeOrToken.AsNode())
+                .OfType<TStatementSyntax>()
+                .SkipWhile(node => node != declarationSyntax)
+                .ToArray();
+
+            for (var statementIndex = 0; statementIndex < statementsAfterDeclaration.Length; statementIndex++)
             {
-                var node = (TStatementSyntax)nodeOrToken.AsNode();
-                if (node is null)
+                var node = statementsAfterDeclaration[statementIndex];
+                if (ContainsReference(node, localVariables, semanticModel, syntaxFactsService, cancellationToken, out var referencedVariable))
                 {
-                    continue;
+                    // Update the statement index this variable was last used in.
+                    lastVariableUsage[referencedVariable] = statementIndex;
                 }
 
-                if (node == declarationSyntax)
+                if (node is TLocalDeclarationSyntax localDeclarationSyntax)
+                {
+                    // Get all symbols declared in this statement.
+                    var operation = semanticModel.GetOperation(localDeclarationSyntax, cancellationToken) as IVariableDeclarationGroupOperation;
+                    var localSymbols = operation.Declarations
+                        .SelectMany(declaration => declaration.Declarators)
+                        .Select(declarator => declarator.Symbol);
+
+                    // Set their declaration index.
+                    foreach (var localSymbol in localSymbols)
+                    {
+                        firstVariableUsage[localSymbol] = statementIndex;
+                    }
+
+                    // Add the variables to the list of locals.
+                    localVariables.AddRange(localSymbols);
+                }
+            }
+
+            // If the trigger variable has no other usage, then return null.
```

Comment:
"then return null" means nothing to me here.  It would be better to state something like "return null to indicate that there is no following statement we need to pull into the using block" (or something like that).